stdlib/asm/crypto/stark.masm

```diff
@@ -0,0 +1,2153 @@
+    dup.1
+    dup.1
+    neq
+    while.true
+        swap.2
+        add.1
+        swap.2
+        #=> [2^cnt, a, cnt, ..]
+        mul.2
+        dup.1
+        dup.1
+        neq
+    end
+    push.0
+    swap.3
+    swap
+    push.0
+    swap.3
+    drop
+    #=> [a, log2(a), y, y]
+end
+
+#================================================
+#   Memory layout:
+#   0. Seed
+#   1. [lde_size, log2(lde_size), 0, 0]
+#   2. Main trace commitment
+#   3. Aux trace commitment
+#   4. Constraint polynomial columns trace commitment
+#   5. [lde_size, log2(lde_size), lde_domain_generator, 0]
+#   6. [num_fri_layers, remainder_size, lde_size, lde_domain_generator]
+#   7. Remainder commitment
+#   8. Intermidiate values.
+#   9. [lde_size, log2(lde_size), lde_domain_generator, 0] (same as 5 but for a different purpose)
+#   10.[gz1, gz0, z1, z0]
+#   11.[x, _, _, _] where x := offset * lde_domain_generator^query_index
+#   20. [0, 0, 0, 1] used for hmerge in case the number of elements to be hashed is not divisible by 8
+#   21. [0, 0, 0, 0] used for hmerge in case the number of elements to be hashed is divisible by 8
+#   22 -> 1999 Empty for now but should hold the public inputs in the future
+#   2000 -> 2015    auxiliary trace random coefficients
+#   2016 -> 2250    constraint composition coefficients
+#   2251    [(z^8)_1, (z^8)_0, z_1, z_0] where z = (z_0, z_1) is the OOD random element
+#   2252 -> 2292    current row of the OOD evaluation frame 
+#   2293 -> 2333    next row of the OOD evaluation frame 
+#   2334 -> 2337    value_i for i in 0..=7 where H(z) := \sum_0^7 value_i * z^i and H is the constraint compostion polynomial
+#   2338 -> 2463    Deep query coefficients
+#   2473 -> 2499    [0, query, evaluation_1, evaluation_0] for the 27 FRI queries
+#   2500 -> ptr_remainder - 1 where ptr_remainder := 2500 + 2 * number_of_layers
+#   ptr_remainder -> ptr_remainder + remainder_length / 2
+#   ptr_remainder + remainder_length / 2 + 1 -> ptr_remainder + remainder_length / 2 + length_of_remainder_polynomial/2
+#================================================
+
+# Notes: 
+# 1. The choice of the number of locals is based on an estimate and should be updated  
+# 2. In particular, once the conjugate term in the deep queries has been removed through
+#   the new Winterfell release, the number of deep query coefficients should be reduced by a 1/4.
+#    The current MASM implementation assumes this when computing the sums but not for the number of 
+#    coefficients. This should be updated and aligned.
+# 3. Once `RandomCoin` has been redesigned, the way randomness is generated should be simplified.
+# 4. The periodic values for the OOD evaluation frame are not implemented. This should probably be
+#   delegated to the constraint evaluator.  
+# 5. Unless otherwise stated, `x`, `X`, `y` and `Y` denote "garbage" values on the stack. When computing
+#   the deep evaluation queries, `x` will denote `offset * g^index`.  
+export.verify.4222
```

Comment:
This probably needs to be updated in a couple of ways:
1. Would be good to spell out what are the expected input/output stack states. For input, I'm guessing it should be trace length and hash of public inputs. For outputs, I think the stack state should be empty.
  a. In the future, we could make the number of queries a public input as well - though, for now I think it is fixed at 27.
2. I think the notes section should be updated as it refers to updating things for the new Winterfell release (which I believe has already been done).

Reply:
1. That's correct. The input stack would be the hash of the public inputs plus `log(trace_length)`. The number of queries could also become a stack-input when we decide to make it variable instead of fixed to `27`.
2. Yes, I did update them to remove outdated notes.

Reply:
One more remark, currently the public inputs are not unhashed but only used to generate the seed for the random coin. This is why the code takes the program hash as stack-input so as to serve as a reminder for the changes that need to be mase.
I think it makes sense to create a separate issue just for dealing with public inputs. What do you think?